src/jit/stacklevelsetter.cpp

```diff
@@ -0,0 +1,266 @@
+            unsigned          numSlots = putArgNumSlots[putArg];
+            SubStackLevel(numSlots);
+        }
+
+#if !FEATURE_FIXED_OUT_ARGS
+        // Set throw blocks incoming stack depth for x86.
+        bool useThrowHlpBlk = !comp->opts.compDbgCode;
+        bool operMightThrow = ((node->gtFlags & GTF_EXCEPT) != 0);
+        if (!framePointerRequired && useThrowHlpBlk && operMightThrow)
+        {
+            SetThrowHelperBlocks(node, block);
+        }
+#endif // !FEATURE_FIXED_OUT_ARGS
+
+        if (node->IsCall())
+        {
+            GenTreeCall* call = node->AsCall();
+
+            unsigned usedStackSlotsCount = PopArgumentsFromCall(call);
+#if defined(UNIX_X86_ABI)
+            assert(call->fgArgInfo->GetStkSizeBytes() == usedStackSlotsCount * TARGET_POINTER_SIZE);
+            call->fgArgInfo->SetStkSizeBytes(usedStackSlotsCount * TARGET_POINTER_SIZE);
+#endif // UNIX_X86_ABI
+        }
+    }
+    assert(currentStackLevel == 0);
+}
+
+#if !FEATURE_FIXED_OUT_ARGS
+//------------------------------------------------------------------------
+// SetThrowHelperBlocks: Set throw helper blocks incoming stack levels targeted
+//                       from the node.
+//
+// Notes:
+//   one node can target several helper blocks.
+//
+// Arguments:
+//   node - the node to process;
+//   block - the source block for the node.
+void StackLevelSetter::SetThrowHelperBlocks(GenTree* node, BasicBlock* block)
+{
+    // Check that it uses throw block, find its kind, find the block, set level.
+    switch (node->OperGet())
+    {
+        case GT_ARR_BOUNDS_CHECK:
+#ifdef FEATURE_SIMD
+        case GT_SIMD_CHK:
+#endif // FEATURE_SIMD
+        {
+            GenTreeBoundsChk* bndsChk = node->AsBoundsChk();
+            SetThrowHelperBlock(bndsChk->gtThrowKind, block);
+        }
+        break;
+
+        case GT_CKFINITE:
+        {
+            SetThrowHelperBlock(SCK_ARITH_EXCPN, block);
+        }
+        break;
+    }
+    if (node->gtOverflowEx())
+    {
+        SetThrowHelperBlock(SCK_OVERFLOW, block);
+    }
+}
+
+//------------------------------------------------------------------------
+// SetThrowHelperBlock: Set throw helper block incoming stack levels targeted
+//                      from the block with this kind.
+//
+// Arguments:
+//   kind - the special throw-helper kind;
+//   block - the source block that targets helper.
+void StackLevelSetter::SetThrowHelperBlock(SpecialCodeKind kind, BasicBlock* block)
+{
+    Compiler::AddCodeDsc* add = comp->fgFindExcptnTarget(kind, comp->bbThrowIndex(block));
+    assert(add != nullptr);
+    if (add->acdStkLvlInit)
+    {
+        if (add->acdStkLvl != currentStackLevel)
+        {
+            framePointerRequired = true;
+            return;
+        }
+    }
+    else
+    {
+        add->acdStkLvlInit = true;
+        if (add->acdStkLvl != currentStackLevel && !comp->codeGen->isFramePointerRequired())
```

Comment:
Isn't `comp->codeGen->isFramePointerRequired()` always false? Otherwise, why are we bothering to do any stack setting?

Reply:
Fixed